[SPARK-12177] [STREAMING] Update KafkaDStreams to new Kafka 0.9 Consumer API (Closes #10681)

markgrover opened the pull request:
First off, many thanks and credit to @nikit-os for starting this work. I am building up on his good work done at #10681.

Here is some context: Kafka has a new consumer API starting Apache Kafka 0.9 (Kafka 0.9.0.0, to be precise). This API supports security features like authentication (via Kerberos) and encryption (via SSL). This new API is in beta but is the only way we can integrate Spark and Kafka in a secure environment.

Kafka 0.9 ships with both the new and old consumer API. So, we can up our dependency of Kafka from 0.8\* to 0.9.  However, the old consumer API in Kafka 0.9 is binary incompatible with the old consumer API in Kafka 0.8_. There are also some minor API compatibility issues between the old API of Kafka 0.8._ and Kafka 0.9, mostly around ZkClient and ZkUtils API. In almost all cases, when users are using the old API whether they are using Kafka 0.8 or Kafka 0.9 with Spark, they won't have to modify any code on their side. It's only when they use the _new_ consumer API from Kafka 0.9, they will have to modify their code.

In addition to supporting to the new API from Kafka 0.9, we can support the old API from Kafka 0.8 only (which works with Kafka 0.8 and Kafka 0.9 brokers), or old API from Kafka 0.9 (which only works with Kafka 0.9 brokers). If we support the old API from 0.8.0, we complicate our integration. We will have to build 2 artifacts (kafka-assembly jars, that is), one for Kafka 0.8 and Kafka 0.9. We'll have to do maven profiles, one for Kafka 0.8 and one for 0.9. Alternatively, if we only support old API from 0.9, users will have upgrade Kafka brokers to 0.9, when using Spark 2.0.

In fact, users should ideally upgrade Kafka brokers before they upgrade Spark since old 0.8.0 clients can work with the new 0.9.0 brokers). I am personally ok with requiring users to upgrade Kafka brokers to 0.9.0 since 1) this is going in Spark 2.0, 2) that's what the Kafka community is pushing for, 3) I think the baggage of supporting Kafka 0.8's old consumer API might be too much to carry until another major release of Spark. Again, with Kafka 0.9's old consumer API, which we support, no user code needs to be changed, they do need to upgrade their brokers though.

As far as the code in this PR goes, we decided to put code for the new API in a new package (called org.apache.spark.streaming.kafka.newapi*), creating a separate maven artifact spark-streaming-kafka-newapi_2.10.

In essence, this means that there are two KafkaCluster classes and two KafkaUtils classes, etc. (similar to how we have NewHadoopRDD) because there wasn't enough duplication between the two implementations. Where there was enough duplication, example KafkaTestUtils, there is only one implementation.

I'd love to get this reviewed. Thanks in advance!

The examples of the new API have some extra required parameters, I will change it shortly for them to have reasonable defaults and be optional. 

Follow on work:
- This is only scala/java implementation. I am working on and currently testing a python implementation in a separate branch at https://github.com/markgrover/spark/tree/kafka09-integration-python I want to get this PR out and committed first. It'll also help me gather feedback sooner than later.
- Add and test authentication with kerberos. For that, we need delegation tokens in Kafka (which is currently unresolved: https://issues.apache.org/jira/browse/KAFKA-1696). This is because, especially in the case of DirectStream, all executors can consume from Kafka brokers and initializing credentials at each executor can been seen as a DoS attack by the Ticket Granting server.
- Encryption exists in this patch but I haven't spent much time integration testing it. If folks prefer, taking that out until I have gotten a chance to integration test it on my own cluster, I am open to that. It's only a few lines and a test that needs to be taken out.

markgrover committed:
Getting rid of an extra dependency

nikit-os committed:
[SPARK-12177] [STREAMING] Update KafkaDStreams to new Kafka 0.9 Consumer API

nikit-os committed:
Merge pull request #1 from markgrover/kafka09-integration

Getting rid of an extra dependency

nikit-os committed:
Fix Kafka params in Java example

markgrover committed:
Removing Broker.scala since it's not needed

markgrover committed:
Scala style related changes to not exceed the line length

markgrover committed:
Formatting changes for styling, also updating scaladoc for functions.

markgrover committed:
Adding missing leaderhost parameter to various methods in OffsetRange.scala

markgrover committed:
Fixing a compilation failure related to previous commit

nikit-os committed:
Merge pull request #2 from markgrover/kafka09-integration

More changes

markgrover committed:
Refactoring Java/Scala code to have use the package name newapi instead of v09 to avoid confusion. Python work is being done in a separate branch github.com/markgrover/spark/tree/kafka09-integration-python

markgrover committed:
Fixing a scalastyle error

markgrover committed:
Merge branch 'master' into kafka09-integration

markgrover committed:
Fixing scalastyle errors related to import ordering

markgrover committed:
Fixing a compilation error

markgrover committed:
More import ordering fixes

markgrover committed:
Function declaration formatting in KafkaUtils and other minor formatting changes

markgrover committed:
Removing support for Kafka 0.8 completely. You can use the new api or the old api using 0.9.0.0 only.

markgrover committed:
Making the new api KafkaTestUtils to not use zkClient explicitly

markgrover committed:
Removing the separate copy of KafkaTestUtils

markgrover commented:
@tdas @harishreedharan @koeninger would appreciate your review here. Thanks!

SparkQA commented:
**[Test build #50220 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50220/consoleFull)** for PR 10953 at commit [`b35ea65`](https://github.com/apache/spark/commit/b35ea6500ca6ea9024632476e266bef2828fb23d).

SparkQA commented:
**[Test build #50220 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50220/consoleFull)** for PR 10953 at commit [`b35ea65`](https://github.com/apache/spark/commit/b35ea6500ca6ea9024632476e266bef2828fb23d).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/50220/
Test FAILed.

markgrover committed:
Fixing a scalastyle error

SparkQA commented:
**[Test build #50224 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50224/consoleFull)** for PR 10953 at commit [`b3f163e`](https://github.com/apache/spark/commit/b3f163e096aa971ccb022773058c7716410299c8).

SparkQA commented:
**[Test build #50224 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50224/consoleFull)** for PR 10953 at commit [`b3f163e`](https://github.com/apache/spark/commit/b3f163e096aa971ccb022773058c7716410299c8).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/50224/
Test FAILed.

markgrover commented:
OK, I have no idea what Mima is but I will take a look and try to run them locally and fix the issues.

vanzin commented:
MiMA is a binary compatibility checker. It's complaining that some changes you made caused the public APIs exposed in the compiled classes to change - meaning existing code compiled against the current Spark version might not run on the next Spark.

First I'd look at whether those changes are necessary; if they are, it might be ok to add exclusions because we're being a bit lenient with API breakages in 2.0.

markgrover committed:
Adding excludes for binary compat testing. The changes were for 2.0 and to a private member so excludes are ok

SparkQA commented:
**[Test build #50332 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50332/consoleFull)** for PR 10953 at commit [`41a8816`](https://github.com/apache/spark/commit/41a8816bb971681b1c69db3c04dddd10ab3c1370).

markgrover commented:
Thanks Marcelo, I added excludes because I think it's ok to have such minor binary incompatibility in this case (Kafka 0.8->Kafka 0.9, Spark 1.6->Spark 2.0).

SparkQA commented:
**[Test build #50332 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50332/consoleFull)** for PR 10953 at commit [`41a8816`](https://github.com/apache/spark/commit/41a8816bb971681b1c69db3c04dddd10ab3c1370).
 * This patch passes all tests.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/50332/
Test PASSed.

mariobriggs commented on external/kafka-newapi/src/main/scala/org/apache/spark/streaming/kafka/newapi/OffsetRange.scala:
```diff
@@ -0,0 +1,133 @@
+
+package org.apache.spark.streaming.kafka.newapi
+
+import kafka.common.TopicAndPartition
```
This import has to be deleted.

mariobriggs commented on external/kafka-newapi/src/main/scala/org/apache/spark/streaming/kafka/newapi/OffsetRange.scala:
```diff
@@ -0,0 +1,133 @@
+  }
+
+  /** Kafka TopicAndPartition object, for convenience */
+  def topicAndPartition(): TopicAndPartition = TopicAndPartition(topic, partition)
```
This has to be 'TopicPartition' and not 'TopicAndPartition'

mariobriggs commented on external/kafka-newapi/src/main/scala/org/apache/spark/streaming/kafka/newapi/OffsetRange.scala:
```diff
@@ -0,0 +1,133 @@
+    new OffsetRange(topic, partition, fromOffset, untilOffset)
+
+  def create(
+      topicAndPartition: TopicAndPartition,
```
this line has to be 'topicPartition: TopicPartition,'

mariobriggs commented on external/kafka-newapi/src/main/scala/org/apache/spark/streaming/kafka/newapi/KafkaRDD.scala:
```diff
@@ -0,0 +1,237 @@
+    messageHandler: ConsumerRecord[K, V] => R
+  ) extends RDD[R](sc, Nil) with Logging with HasOffsetRanges {
+
+  private val KAFKA_DEFAULT_POLL_TIME: String = "0"
```
this default has to be something greater than 0. Since KafkaRDDIterator.getNext() is implemented as a recursive function below, we run into a StackOverflow exception when this is 0. 

This also means if the user specifically also sets this to 0 via 'spark.kafka.poll.time', do we need to guard against it? (or move to a simple imperative for loop)

markgrover replied:
Hi Mario, Based on these API docs, it seems like 0 is a valid option:
https://kafka.apache.org/090/javadoc/org/apache/kafka/clients/consumer/KafkaConsumer.html#poll(long)

If you can explain further why 0 is problematic, I'd appreciate that, and am happy to make suggested changes, I don't see the problem just yet:-)

mariobriggs replied:
U can reproduce the problem by running the test suites without setting this - "spark.kafka.poll.time" -> "1000".

So problem basically is that we had implemented the situation when timeout is too small (i.e. poll returns no data), as a recursive function call. So for e.g. if it actually takes 700 ms for the kafkaClient to get the data from the kakfaServer, then for that period of 700 ms, we are making continous recursive function calls which cause a JVM stackoverflow.

I have issued a PR to your repo, where i reimplemented this functionality using a imperative loop rather than a recursive function call

mariobriggs commented on external/kafka-assembly/pom.xml:
```diff
@@ -43,6 +43,11 @@
+      <version>${project.version}</version>
+    </dependency>
+    <dependency>
+      <groupId>org.apache.spark</groupId>
```
Mark, the jar  (spark-streaming-kafka-assembly*.jar)  produced by this assembly works only with kafka 0.9 brokers/servers. I have tested it.

So if i want to use Spark 2.0,  and continue to run my existing spark-streaming-kakfka-app (uses 'older' spark-streaming-kakfa api) because i cannot upgrade  kafka brokers/servers to 0.9, there is no spark-streaming-kafka-assembly jar provided by spark 2.0 for that?

I think we need to provide for above, which means we need produce a 'spark-streaming-kafka-newapi-assembly*.jar' and a 'spark-streaming-kafka-assembly*.jar' .  I am not sure if it is a best practice to produce 2 diff assembly.jar from a single pom.xml if there is a way, else just have a spark-kafka-assembly-newapi

mariobriggs replied:
Mark, i re-read your description above and i think you are suggesting that we support old consumer API's only with 0.9 brokers, in which case we dont need the above.

mariobriggs commented on external/kafka-newapi/src/main/scala/org/apache/spark/streaming/kafka/newapi/KafkaUtils.scala:
```diff
@@ -0,0 +1,601 @@
+  def createOffsetRange(topic: String, partition: JInt, fromOffset: JLong, untilOffset: JLong):
+    OffsetRange = OffsetRange.create(topic, partition, fromOffset, untilOffset)
+
+  def createTopicAndPartition(topic: String, partition: JInt): TopicAndPartition =
```
this method and the import of TopicAndPartition has to be knocked off

mariobriggs commented on external/kafka-newapi/src/main/scala/org/apache/spark/streaming/kafka/newapi/KafkaCluster.scala:
```diff
@@ -0,0 +1,184 @@
+    getOffsetsWithoutLeaders(topicPartitions, OffsetResetStrategy.EARLIEST)
+  }
+
+  def getPartitions(topics: Set[String]): Set[TopicPartition] = {
```
nitpick: this should be implemented by reusing getPartitionsInfo() like below

```  
getPartitionInfo(topics).map { pi =>
      new TopicPartition(pi.topic, pi.partition)
    }
```

markgrover committed:
Implementing some of the review feedback from Mario

markgrover committed:
More feedback implemented from Mario

markgrover committed:
Getting rid of some duplicate code

SparkQA commented:
**[Test build #50426 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50426/consoleFull)** for PR 10953 at commit [`82223f8`](https://github.com/apache/spark/commit/82223f8047ae5b7b0b3df0efb4661e792f5d8b5c).

SparkQA commented:
**[Test build #50426 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50426/consoleFull)** for PR 10953 at commit [`82223f8`](https://github.com/apache/spark/commit/82223f8047ae5b7b0b3df0efb4661e792f5d8b5c).
 * This patch passes all tests.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/50426/
Test PASSed.

markgrover committed:
Merge branch 'master' into kafka09-integration

markgrover committed:
Getting rid of extra whitespace

markgrover commented:
@mariobriggs, thanks for your feedback. I have incorporated all but one of it (related to poll time). In that case, I have asked a follow up quesiton. If I missed something else, please do let me know.

@tdas @koeninger @harishreedharan, I'd still really appreciate your review of this PR.

markgrover committed:
Merge branch 'master' into kafka09-integration

markgrover committed:
Updating the new kafka-newapi module to be 2.11 based as well

mariobriggs committed:
Update KafkaRDD.scala

implement fetchBatch() as a loop when kafka.poll is not large enough interval for data to reach kafkaClient from kafkaServer

SparkQA commented:
**[Test build #50568 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50568/consoleFull)** for PR 10953 at commit [`d932d75`](https://github.com/apache/spark/commit/d932d75f520866169797a311ff64939eda7ef1c6).

markgrover committed:
Correcting the package name in the wordcount example

markgrover committed:
Getting rid of extra parameter of batch interval in examples. Parameters, especially required ones should be few so it's very simple to run the example

SparkQA commented:
**[Test build #50568 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50568/consoleFull)** for PR 10953 at commit [`d932d75`](https://github.com/apache/spark/commit/d932d75f520866169797a311ff64939eda7ef1c6).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/50568/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/50574/
Test FAILed.

markgrover commented:
The previous test failures seems like a timeout when fetching the PR from github:
```
> git fetch --tags --progress https://github.com/apache/spark.git +refs/pull/10953/*:refs/remotes/origin/pr/10953/* # timeout=15
ERROR: Timeout after 15 minutes
ERROR: Error fetching remote repo 'origin'
```

markgrover commented:
Jenkins, retest this please.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/50577/
Test FAILed.

markgrover commented:
Can someone add me to the whitelist please? Thanks!

vanzin commented:
You already are on it, given the tests for you PR started automatically when you submitted it.

markgrover commented:
Ok, thanks, Marcelo. I didn't think I was because Jenkins didn't trigger on my explicit request but I will try again.

markgrover commented:
Jenkins, retest this please.

SparkQA commented:
**[Test build #50588 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50588/consoleFull)** for PR 10953 at commit [`1f77501`](https://github.com/apache/spark/commit/1f77501c5bee5f35d7a87126f33740de48c9a47a).

SparkQA commented:
**[Test build #50588 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50588/consoleFull)** for PR 10953 at commit [`1f77501`](https://github.com/apache/spark/commit/1f77501c5bee5f35d7a87126f33740de48c9a47a).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/50588/
Test PASSed.

markgrover committed:
Adding a change that will conflict with Mario's change for now

markgrover committed:
Merge pull request #2 from mariobriggs/patch-1

Avoid StackOverflowException when Kafka.poll() 'timeout' param is 0

markgrover committed:
Changes related to poll time.
Kafka Console Consumer seems to be use Long.MaxValue so that's what is best for us to use

SparkQA commented:
**[Test build #50609 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50609/consoleFull)** for PR 10953 at commit [`8fae471`](https://github.com/apache/spark/commit/8fae47170a6fe1f364b77abeac1673d315de0bad).

SparkQA commented:
**[Test build #50609 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50609/consoleFull)** for PR 10953 at commit [`8fae471`](https://github.com/apache/spark/commit/8fae47170a6fe1f364b77abeac1673d315de0bad).
 * This patch **fails to build**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/50609/
Test FAILed.

markgrover commented:
Jenkins, test this please.

SparkQA commented:
**[Test build #50613 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50613/consoleFull)** for PR 10953 at commit [`7dcb848`](https://github.com/apache/spark/commit/7dcb8482e0daac698380650e1b3f2ce4e85e3274).

SparkQA commented:
**[Test build #50613 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50613/consoleFull)** for PR 10953 at commit [`7dcb848`](https://github.com/apache/spark/commit/7dcb8482e0daac698380650e1b3f2ce4e85e3274).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/50613/
Test FAILed.

SparkQA commented:
**[Test build #50610 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50610/consoleFull)** for PR 10953 at commit [`7dcb848`](https://github.com/apache/spark/commit/7dcb8482e0daac698380650e1b3f2ce4e85e3274).

SparkQA commented:
**[Test build #50610 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50610/consoleFull)** for PR 10953 at commit [`7dcb848`](https://github.com/apache/spark/commit/7dcb8482e0daac698380650e1b3f2ce4e85e3274).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/50610/
Test FAILed.

markgrover committed:
Updating the import ordering

SparkQA commented:
**[Test build #50615 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50615/consoleFull)** for PR 10953 at commit [`edfca9f`](https://github.com/apache/spark/commit/edfca9f933073d5d32ee1e07536f5ca80c53bcae).

markgrover committed:
Fixing up some minor things in the Example

markgrover committed:
Correcting comment

SparkQA commented:
**[Test build #50615 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50615/consoleFull)** for PR 10953 at commit [`edfca9f`](https://github.com/apache/spark/commit/edfca9f933073d5d32ee1e07536f5ca80c53bcae).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/50615/
Test FAILed.

markgrover committed:
Reintroducing if loop because some times buffer may return empty

harishreedharan commented on external/kafka-newapi/pom.xml:
```diff
@@ -0,0 +1,122 @@
+  </parent>
+
+  <groupId>org.apache.spark</groupId>
+  <artifactId>spark-streaming-kafka-newapi_2.11</artifactId>
```
Is it possible to name this module something else? `newapi` is pretty generic - and really refers to kafka's new API while the name seems to suggest it is some new API is Spark

harishreedharan commented on external/kafka-newapi/src/main/scala/org/apache/spark/streaming/kafka/newapi/DirectKafkaInputDStream.scala:
```diff
@@ -0,0 +1,208 @@
+  override protected[streaming] val rateController: Option[RateController] = {
+    if (RateController.isBackPressureEnabled(ssc.conf)) {
+      Some(new DirectKafkaRateController(id,
+        RateEstimator.create(ssc.conf, ssc_.graph.batchDuration)))
```
Why use both `ssc` and `ssc_`?

harishreedharan commented:
I have a more fundamental question - given that this patch does not add a whole lot of new functionality but just ports the currently functionality to use the new Kafka API, is there a reason we can't refactor the current `KafkaRDD` and `*KafkaDStream` classes such that the functionality accessing the Kafka API is wrapped in methods(say `fetchFromKafka` or something), but everything else (rate limiting, checkpointing, producing new RDDs) etc is "outside". That way the new API-use can be simply in overridden methods (`override def fetchFromKafka`). 

Also, can we change the package name and module names to something other than `newapi` - I mean `new` is relative - Kafka could end up having a newer API at some point.

markgrover committed:
WIP regarding getting rid of spark.kafka.poll.time in Kafka Params

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/50709/
Test FAILed.

markgrover commented:
Jenkins, retest this please.

SparkQA commented:
**[Test build #50718 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50718/consoleFull)** for PR 10953 at commit [`10a18b8`](https://github.com/apache/spark/commit/10a18b8482eeb21ce69c2c7fbbffc8b3b0932398).

SparkQA commented:
**[Test build #50718 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50718/consoleFull)** for PR 10953 at commit [`10a18b8`](https://github.com/apache/spark/commit/10a18b8482eeb21ce69c2c7fbbffc8b3b0932398).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/50718/
Test PASSed.

markgrover committed:
WIP refactor of old KafkaRDD

markgrover committed:
With this change, the only files that still need to be taken care of are KafkaUtils and DirectKafkaInputDStream. KafkaRDD and all its dependencies have been taken care of.

markgrover committed:
Everything seems to be refactored except KafkaUtils and Java code

markgrover committed:
Refactoring of Kafka Utils. Left is java code, test code, and examples, etc.

markgrover committed:
Complete refactor of KafkaUtils for Scala and potentially Java and changed example

markgrover committed:
Refactor of DirectKafkaStreamSuite

markgrover committed:
Taken care of KafkaClusterSuite

markgrover committed:
KafkaRDDSuite is done too. Now Java code and Java tests are the only ones left

markgrover committed:
Fixing examples pom file to not have a reference to the old package name

markgrover committed:
Adding Java tests to the original place

markgrover committed:
Adding the Java example. It should compile but it doesn't seem to be doing so.

markgrover committed:
Fixing some test bugs and adding some improvements related to clean up

markgrover committed:
Selective revert of the previous patch since the randoms weren't needed, I think

markgrover committed:
Test improvements. Now we are only down to only the new tests failing

markgrover committed:
Fixing scala style but also minor bug fix in OffsetRanges and polltime

markgrover committed:
Finally figured out how to read the configuration variable from within Spark

markgrover committed:
Adding ignore for a test that was failing. This is because of a kafka issue - KAFKA-3029 issue.

markgrover committed:
Adding some extra functions in OffsetRanges

markgrover committed:
Adding excludes for binary compatibility

markgrover committed:
Getting rid of unnecessary getPartitions method in the base class, making it abstract

markgrover committed:
Fixing a Java test bug

markgrover committed:
WIP - offset recovery test is failing but this is testing to have no change in parameters compared to the previous API implementation

markgrover committed:
WIP - offset recovery test is failing but this is testing to have no change in parameters compared to the previous API implementation

markgrover committed:
Removing all the extra separate subproject code

markgrover commented:
Thanks for your feedback, @harishreedharan I have done a significant refactoring of code based on your feedback, sharing code where possible. The new RDD, is now called NewKafkaRDD (similar to how Hadoop RDD was done) and similar naming convention for NewKafkaCluster and NewDirectKafkaInputDStream, etc.

[One of the tests](https://github.com/markgrover/spark/blob/kafka09-integration/external/kafka/src/test/scala/org/apache/spark/streaming/kafka/DirectKafkaStreamSuite.scala#L381) is currently marked as ignored because it's blocked by KAFKA-3029. Once that's committed and available in Apache Kafka 0.9.0.1, I will re-enable that test.

Once again, thanks for your review. And, I'd appreciate a review from others as well.

SparkQA commented:
**[Test build #50916 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50916/consoleFull)** for PR 10953 at commit [`abd167c`](https://github.com/apache/spark/commit/abd167c89c448261b70a7e7b1a755d9df21cd74a).

SparkQA commented:
**[Test build #50916 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50916/consoleFull)** for PR 10953 at commit [`abd167c`](https://github.com/apache/spark/commit/abd167c89c448261b70a7e7b1a755d9df21cd74a).
 * This patch passes all tests.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/50916/
Test PASSed.

vgromakowski commented:
Authentication and access control are mandatory for enterprise deployments with security constraints, specially multi tenant ones. I have lots of interest for this PR !

mariobriggs commented on external/kafka/src/main/scala/org/apache/spark/streaming/kafka/DirectKafkaInputDStreamBase.scala:
```diff
@@ -0,0 +1,158 @@
+ *  starting point of the stream
+ */
+private[streaming]
+abstract class DirectKafkaInputDStreamBase[
```
thinking another way. Instead of having a common base class and then 2 classes that extend it, could we introduce a 'BackPressure' trait that is then mixed-in 

Benefits :  we don't have to force fit things like Decoder's in base class  type (especially given we are force fitting the deprecated thing in base)

We still might not be able to do away with the TopicPartition vs TopicAndPartition even when using the BackPressure trait. But in those scenario's, i think we should use the new consumer class as a rule of thumb rather than the old kafka.xxx classes (after all that is what we will knock off over period of time and not the other way around and also we force the old impl to do the extra conversion step and not the new impl.... few nanosecond gain :-) )

markgrover replied:
Hi Mario, thanks for your feedback. This is a reasonable request. However, I think both you and I realize this is relatively minor and see my comment below about the new API not being the go to recommendation just yet. So, I will keep this on the back burner for now.

mariobriggs commented on external/kafka/src/main/scala/org/apache/spark/streaming/kafka/KafkaUtils.scala:
```diff
@@ -610,6 +619,417 @@ object KafkaUtils {
+    * @tparam V type of Kafka message value
+    * @return RDD of (Kafka message key, Kafka message value)
+    */
+  def createNewRDD[K: ClassTag, V: ClassTag](
```
I think we should avoid method names like createNewRDD and createNewDirectStream in the public API. Hari commented about the word 'new' and i think this takes it 1 step further in the -ve direction.

Can we just drop the 'new' here. Anyhow the signatures dont conflict and i assume we will mark the prev ones as deprecated, which is good enough to bring to the user notice as to which ones is expected to be used.

markgrover replied:
Mario, see my comment below about new. As far as marking the previous ones as deprecated, I don't quite feel the same way. The new consumer API is brand new, and may take a bit to bake in (given issues like [KAFKA-3029](issues.apache.org/jira/browse/KAFKA-3029) and [KAFKA-3159](https://issues.apache.org/jira/browse/KAFKA-3159) are being found). It's also currently marked as beta. So, my recommendation would not be to deprecate the old API. It's only for those who are eager and curious to use the new API, and especially the associated encryption and authentication capabilities it offers, for now, in my opinion. Let me know if you differ though.

mariobriggs commented on external/kafka/src/main/scala/org/apache/spark/streaming/kafka/NewDirectKafkaInputDStream.scala:
```diff
@@ -0,0 +1,134 @@
+/*
```
Back to Hari's comment about the word 'new'. On the JIRA we discussed and ruled out using version numbers. It might have been easier if like 'HighLevelAPI/LowLevelAPI/SimpleConsumer', there was a word with a functional element to it that we could use here for the 'newConsumer'

markgrover replied:
Hi Mario, I am following what Hadoop did when it was doing something similar. See NewHadoopRDD, for example. I am actually pretty flexible about the name so if folks disagree and have a better name to suggest, I am totally down for that. I just don't want to rename over and over again, so I'd suggest keeping the name the way it is right now, and get a consensus on what, if any, the new name should be. Once we have it here, I can change it.

mariobriggs commented on external/kafka/src/main/scala/org/apache/spark/streaming/kafka/NewKafkaCluster.scala:
```diff
@@ -0,0 +1,156 @@
+ *                    NOT zookeeper servers, specified in host1:port1,host2:port2 form
+ */
+private[spark]
+class NewKafkaCluster[K: ClassTag, V: ClassTag](val kafkaParams: Map[String, String])
```
This is another place where we can't use 'new' in the class name . See https://github.com/apache/spark/pull/9007 which recently got merged. We have to make this a public class because of valid use-cases like https://issues.apache.org/jira/browse/SPARK-13106

markgrover replied:
Thanks, it did conflict with this PR but I have merged it. I will make the NewKafkaCluster class public as well, similar to this change now.

mariobriggs commented on external/kafka/src/main/scala/org/apache/spark/streaming/kafka/KafkaRDDBase.scala:
```diff
@@ -0,0 +1,105 @@
+      Seq()
+    }
+  }
+
```
any particular reason u moved the errXXX methods here to KafkaUtils ?

markgrover committed:
Merge branch 'master' into kafka09-integration

Conflicts:
	external/kafka/src/main/scala/org/apache/spark/streaming/kafka/KafkaCluster.scala

SparkQA commented:
**[Test build #50946 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50946/consoleFull)** for PR 10953 at commit [`846a85a`](https://github.com/apache/spark/commit/846a85ab1d23176028cc6dabddac505c60db90f2).

markgrover committed:
Merge branch 'kafka09-integration' into kafka09-integration-refactor-params

markgrover committed:
Merge branch 'kafka09-integration-refactor-params' of github.com:markgrover/spark into kafka09-integration-refactor-params

markgrover committed:
Fixing the test by clearing the collection

SparkQA commented:
**[Test build #50946 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/50946/consoleFull)** for PR 10953 at commit [`846a85a`](https://github.com/apache/spark/commit/846a85ab1d23176028cc6dabddac505c60db90f2).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/50946/
Test PASSed.

markgrover committed:
Updating the excludes

markgrover committed:
Making NewKafkaCluster class public

SparkQA commented:
**[Test build #51125 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51125/consoleFull)** for PR 10953 at commit [`9983e7d`](https://github.com/apache/spark/commit/9983e7d07077fbf374811989058bf28eae10cd3a).

SparkQA commented:
**[Test build #51125 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51125/consoleFull)** for PR 10953 at commit [`9983e7d`](https://github.com/apache/spark/commit/9983e7d07077fbf374811989058bf28eae10cd3a).
 * This patch **fails build dependency tests**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51125/
Test FAILed.

markgrover committed:
Merge branch 'master' into kafka09-integration

SparkQA commented:
**[Test build #51193 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51193/consoleFull)** for PR 10953 at commit [`c422cd5`](https://github.com/apache/spark/commit/c422cd5848d29b2995c67788407cf3ba35d0a0a8).

SparkQA commented:
**[Test build #51193 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/51193/consoleFull)** for PR 10953 at commit [`c422cd5`](https://github.com/apache/spark/commit/c422cd5848d29b2995c67788407cf3ba35d0a0a8).
 * This patch **fails build dependency tests**.
 * This patch **does not merge cleanly**.
 * This patch adds the following public classes _(experimental)_:
  * `class KMeansModel(JavaModel, MLWritable, MLReadable):`
  * `class KMeans(JavaEstimator, HasFeaturesCol, HasPredictionCol, HasMaxIter, HasTol, HasSeed,`
  * `class ALS(JavaEstimator, HasCheckpointInterval, HasMaxIter, HasPredictionCol, HasRegParam, HasSeed,`
  * `class ALSModel(JavaModel, MLWritable, MLReadable):`
  * `case class Grouping(child: Expression) extends Expression with Unevaluable `
  * `case class GroupingID(groupByExprs: Seq[Expression]) extends Expression with Unevaluable `
  * `case class AssertNotNull(child: Expression, walkedTypePath: Seq[String])`
  * `case class ReturnAnswer(child: LogicalPlan) extends UnaryNode `
  * `public class UnsafeRowParquetRecordReader extends SpecificParquetRecordReaderBase<InternalRow> `
  * `class ContinuousQueryManager(sqlContext: SQLContext) `
  * `case class CollectLimit(limit: Int, child: SparkPlan) extends UnaryNode `
  * `trait BaseLimit extends UnaryNode `
  * `case class LocalLimit(limit: Int, child: SparkPlan) extends BaseLimit `
  * `case class GlobalLimit(limit: Int, child: SparkPlan) extends BaseLimit `
  * `case class TakeOrderedAndProject(`
  * `class ContinuousQueryListenerBus(sparkListenerBus: LiveListenerBus)`
  * `class FileStreamSource(`
  * `trait HadoopFsRelationProvider extends StreamSourceProvider `
  * `abstract class ContinuousQueryListener `

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/51193/
Test FAILed.

markgrover committed:
Merge branch 'master' into kafka09-integration

Conflicts:
	project/MimaExcludes.scala

SparkQA commented:
**[Test build #52280 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52280/consoleFull)** for PR 10953 at commit [`075226e`](https://github.com/apache/spark/commit/075226e32adbc8fd6b7412518d1e84bb5960a431).

SparkQA commented:
**[Test build #52280 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52280/consoleFull)** for PR 10953 at commit [`075226e`](https://github.com/apache/spark/commit/075226e32adbc8fd6b7412518d1e84bb5960a431).
 * This patch **fails build dependency tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/52280/
Test FAILed.

markgrover commented:
Will fix shortly. I was building against a custom built version of kafka
because 0.9.1 wasn't out which had some blocker fixes, to make tests pass.
On Mar 1, 2016 6:00 PM, "UCB AMPLab" <notifications@github.com> wrote:

> Test FAILed.
> Refer to this link for build results (access rights to CI server needed):
> https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/52280/
> Test FAILed.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/apache/spark/pull/10953#issuecomment-191016757>.
>

markgrover committed:
Taking the SNAPSHOT out since 0.9.0.1 has now been released

SparkQA commented:
**[Test build #52322 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52322/consoleFull)** for PR 10953 at commit [`993c6fd`](https://github.com/apache/spark/commit/993c6fd1b4b26d9fb68ec9c2444d16df4a033cb2).

SparkQA commented:
**[Test build #52322 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52322/consoleFull)** for PR 10953 at commit [`993c6fd`](https://github.com/apache/spark/commit/993c6fd1b4b26d9fb68ec9c2444d16df4a033cb2).
 * This patch **fails PySpark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/52322/
Test FAILed.

markgrover commented:
Hmm, this is weird, these tests were passing before. Anyways, I was able to circumvent them by setting SPARK_DRIVER_MEMORY to a higher value on my local box. I'll kick off another test run to see if the failure was transient. And, if not, I'd have to see if a bump is necessary here too, or why is it necessary now.

markgrover commented:
Jenkins, retest this please.

SparkQA commented:
**[Test build #52352 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52352/consoleFull)** for PR 10953 at commit [`993c6fd`](https://github.com/apache/spark/commit/993c6fd1b4b26d9fb68ec9c2444d16df4a033cb2).

markgrover committed:
Getting rid of TODOs since Kafka 0.9.0.1 is now released

SparkQA commented:
**[Test build #52353 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52353/consoleFull)** for PR 10953 at commit [`229b773`](https://github.com/apache/spark/commit/229b773f9c5c894088a29937f241ad0db48991ba).

SparkQA commented:
**[Test build #52352 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52352/consoleFull)** for PR 10953 at commit [`993c6fd`](https://github.com/apache/spark/commit/993c6fd1b4b26d9fb68ec9c2444d16df4a033cb2).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/52352/
Test FAILed.

SparkQA commented:
**[Test build #52353 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/52353/consoleFull)** for PR 10953 at commit [`229b773`](https://github.com/apache/spark/commit/229b773f9c5c894088a29937f241ad0db48991ba).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/52353/
Test FAILed.

HyukjinKwon commented:
@markgrover Mind adding `Closes #10681` in the PR description so that merging script can close that together?

Xaerxess commented:
Kafka 0.10.0 is out, any plans of merging this PR or updating it with newest Kafka?

koeninger commented:
I think this PR is fairly out of date regarding things like needing to cache consumers on executors, getPreferredLocations, and a separate subproject for kafka-0.8

markgrover commented:
Yeah, I agree with @koeninger. This PR is pretty out of date, it makes sense to turn focus on Cody's PR #11863
